app/inpututils.cpp

```diff
@@ -786,6 +786,26 @@ const QUrl InputUtils::getEditorComponentSource( const QString &widgetName, cons
     return QUrl( path.arg( QLatin1String( "textedit" ) ) );
   }
 
+  if ( widgetName == QStringLiteral( "valuerelation" ) )
+  {
+    QgsMapLayer *referencedLayer = QgsProject::instance()->mapLayer( config.value( "Layer" ).toString() );
+    QgsVectorLayer *layer = qobject_cast<QgsVectorLayer *>( referencedLayer );
+
+    if ( layer )
+    {
+      int featuresCount = layer->allFeatureIds().count();
```

Comment:
This may be inefficient for layers with many features as allFeatureIds() iterates over the whole layer to extract feature IDs. A more efficient approach should be to do layer->dataProvider()->featureCount() - normally the count is already cached within the provider (and typically calculated faster)